that_depends/providers/collections.py

```diff
@@ -18,3 +18,26 @@ def sync_resolve(self) -> list[T]:
 
     async def __call__(self) -> list[T]:
         return await self.async_resolve()
+
+
+class Selector(AbstractProvider[T]):
+    def __init__(self, selector: typing.Callable[[], str], **providers: AbstractProvider[T]) -> None:
+        self._selector = selector
+        self._providers = providers
+
+    async def async_resolve(self) -> T:
+        selected_key = self._selector()
+        if selected_key not in self._providers:
+            msg = f"No provider matches {selected_key}"
+            raise RuntimeError(msg)
+        return await self._providers[selected_key].async_resolve()
+
+    def sync_resolve(self) -> T:
+        selected_key = self._selector()
+        if selected_key not in self._providers:
+            msg = f"No provider matches {selected_key}"
+            raise RuntimeError(msg)
+        return self._providers[selected_key].sync_resolve()
+
+    async def __call__(self) -> T:
+        return await self.async_resolve()
```

Comment:
this method is present in AbstractProvider https://github.com/jasonprado/that-depends/blob/93b36340165cff23b800b46fce616bc227e0ec5c/that_depends/providers/base.py#L10